src/js/node/events.js

```diff
@@ -366,42 +367,56 @@ async function* on(emitter, event, options = {}) {
   const unconsumedErrors = new FixedQueue();
   let done = false;
 
-  emitter.on(event, ev => {
-    if (!unconsumedPromises.isEmpty()) {
-      const { resolve } = unconsumedPromises.shift();
-      return resolve([ev]);
+  const eventHandler = (ev) => {
+    if(!done) {
+      // If there is a pending Promise -> resolve with current event value.
+      if (!unconsumedPromises.isEmpty()) {
+        const { resolve } = unconsumedPromises.shift();
+        return resolve(ev);
+      }
+      // Else: Add event value to queue so it can be consumed by a future Promise.
+      unconsumedEvents.push(ev);
     }
-    unconsumedEvents.push([ev]);
-  });
+  }
+
+  emitter.on(event, options[kFirstEventParam] ? eventHandler : (...args) => eventHandler(args) );
+
```

Comment:
Needed to move the eventHandler logic out of `emitter.on` to handle kFirstEventParam more efficiently. Alternative would be to always use ...args and return args[0] in the case of kFirstEventParam = true, but I remember that those spread operations were quite performance-heavy.

I kinda dislike nested function definitions though, is there a more elegant way to do it?

Reply:
i'm not sure theres much to get around this as the nested function is capturing a lot of shared state.